### PR TITLE
support transmission data pre-worker to post-worker

### PR DESCRIPTION
**Pull request by zhaoyta**
current pre-worker only export WordReport status for judge the flow whether run continue.

but some scene，post worker want to fetch pre worker result .so i modify some code to support post-worker  can fetch pre-worker result.
1, in condition flow , toExecute use pre-worker result. nextOnPredicateSuccess and nextOnPredicateFailure can use toExecute result.
2.in ParallelFlow, each worker use one pre-worker result and produce List<WorkerReport> to post-worker.
3.in sequence flow ,each worker use pre run worker result.but the first use pre-worker result.

### CODE REVIEW

**Comment by mattiasrosberg** [comment]
@benas Is there any progress on this one? 

@zhaoyta , it would be nice if you could re-submit the PR using the same code formatting as the original code base (and perhaps squash the commits into a single one). It would make it easier to follow the changes.

**Comment by benas** [comment]
HI @zhaoyta,

Thank you for opening this PR. Could you please re-submit it without changing code formatting? It is hard to follow your changes at the moment. And please rebase/squash your commits on the latest master.

Thank you upfront!

File: src/main/java/org/jeasy/flows/work/DefaultWorkReport.java [hunk 1]
```diff
@@ -1,71 +1,91 @@
...
  * @author Mahmoud Ben Hassine (mahmoud.benhassine@icloud.com)
+ * @param <T>
  */
-public class DefaultWorkReport implements WorkReport {
+public class DefaultWorkReport<T> implements WorkReport<T> {
+
+  private WorkStatus status;
+  private Throwable error;
+
+  private Map<String, T> collectData;
```
Comment (benas) [inline src/main/java/org/jeasy/flows/work/DefaultWorkReport.java]: Sharing data between work units should be done via the work context and not through the work report . The report may not be ready in a parallel flow. Moreover, this makes the DefaultWorkReport generic which is not needed

File: src/main/java/org/jeasy/flows/work/DefineCallable.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+
+/**
+ * @ClassName: Callable .
+ * @Description:TODO(这里用一句话描述这个类的作用) <\br>
+ * @author: zyt <\br>
+ * @date: 2018年12月13日 下午7:04:36 <\br>
+ * 
+ */
+public interface DefineCallable<T> {
```
Comment (benas) [inline src/main/java/org/jeasy/flows/work/DefineCallable.java]: I see no need to introduce this new interface.

File: src/main/java/org/jeasy/flows/work/Work.java [hunk 3]
```diff
@@ -1,42 +1,40 @@
...
  *
  * Work name must be unique within a workflow.
  */
-public interface Work extends Callable<WorkReport> {
+public interface Work<T> extends DefineCallable<WorkReport<T>> {
 
-    String getName();
+  String getName();
+
+  WorkReport<T> call(List<WorkReport<T>> param, WorkContext context);
```
Comment (benas) [inline src/main/java/org/jeasy/flows/work/Work.java]: Why do we accept a list of reports + the context here? If I implement  this interface for a work that is going to be executed in a parallel flow, the list of reports of previous work units is not ready yet (previous/next does not make sense in a parallel setup).

File: src/main/java/org/jeasy/flows/work/WorkContext.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+package org.jeasy.flows.work;
+
+/**   
+ * @ClassName:  WorkContext   .
+ * @Description:TODO(这里用一句话描述这个类的作用)  <\br> 
+ * @author: zyt <\br>
+ * @date:   2018年12月14日 下午6:31:44   <\br>
+ *   
+ */
+public interface WorkContext {
```
Comment (benas) [inline src/main/java/org/jeasy/flows/work/WorkContext.java]: What is the purpose of this empty interface? IMO the execution context should be a class which contain at least a data structure (like a map) to hold contextual data.

File: src/main/java/org/jeasy/flows/work/WorkReport.java [hunk 5]
```diff
@@ -1,46 +1,48 @@
...
+   * @return execution status
+   */
+  WorkStatus getStatus();
 
-    /**
-     * Get work execution status.
-     * @return execution status
-     */
-    WorkStatus getStatus();
+  Map<String, T> getCollectData();
```
Comment (benas) [inline src/main/java/org/jeasy/flows/work/WorkReport.java]: As mentioned previously, sharing data should be done through the WorkContext and not the WorkReport. What should I return here if my work does not collect any data? I could of course return an empty collection, but this forces me to implement a method that I don't really need, which is a sign that there is something wrong with this design.

**Comment by benas** [comment]
I managed to ignore the formatting noise and was able to spot the major changes.

I agree on the idea of introducing an execution context in general (Here is an example I wrote a couple of years ago https://gist.github.com/benas/4d55e8b18f2ba05ab94a14eda98bd6e9). However, I do not agree on the way this PR implements it. I added some inline comments to show my concerns about the design choices suggested here, but the blocking points are:

* `2.in ParallelFlow, each worker use one pre-worker result and produce List to post-worker.`: Sharing an execution context in a sequential flow makes sense because we can talk about "previous" and "next" work units, what you call "pre"/"post" worker here. However, this does not make sense in a parallel flow (what does "previous"/"next" mean when things are running in parallel?)
* Sharing data between work units should be done via the execution context and not through the work report. We can share a tread-safe context between parallel work units, but we can't pass data via reports as they might not be ready yet (this approach does not work in a parallel setup).
* Introducing a new concept `DefineCallable` which is not needed IMO
* The suggested `WorkContext` interface is empty (and could be a class since it's only a data holder) and the shared data is added to `WorkReport`

@zhaoyta I really appreciate the effort you put in this PR. I tried my best to see how we can improve this PR and  merge it, but I believe we need a completely different approach to implement this feature. Thank you anyway!

**Comment by hzn6426** [comment]
You can refer to the following [Easy Work](https://github.com/hzn6426/easy-work), a similar Java process engine.
